File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_24/java/Solution.java
package com.hi.dhl.algorithms.offer._24.java;

import com.hi.dhl.algorithms.model.ListNode;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/29
 *     desc  :
 * </pre>
 */
class Solution {

    /**
     * 非递归
     */
    public ListNode reverseList(ListNode head) {
        ListNode pre = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = pre;
            pre = head;
            head = next;
        }
        return pre;
    }

    /**
     * 递归
     */
    public ListNode reverseList2(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode cur = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return cur;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_24/kotlin/Solution.kt
package com.hi.dhl.algorithms.offer._24.kotlin

import com.hi.dhl.algorithms.model.ListNode

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/29
 *     desc  :
 * </pre>
 */

class Solution {

    /**
     * 非递归
     */
    fun reverseList(head: ListNode?): ListNode? {
        var pre: ListNode? = null
        var current = head
        while (current != null) {
            val next = current.next
            current.next = pre;
            pre = current;
            current = next
        }
        return pre;
    }

    /**
     * 递归
     */
    fun reverseList2(head: ListNode?): ListNode? {
        if (head == null || head.next == null) return head;
        val cur = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return cur;
    }
}